### PR TITLE
Tcell instrumentation

### PR DESCRIPTION
**Pull request by zyme**
Adds tcell monitoring agent to the backend, similar to other sparkjava projects.  Tarball is kept in GCP bucket, and can be updated as we see fit.

### CODE REVIEW

File: pepper-apis/config/tcell_agent.config.ctmpl [hunk 1]
```diff
@@ -0,0 +1,27 @@
+
+{{with $environment := env "ENVIRONMENT"}}
+{{with $conf := vault (printf "secret/pepper/%s/tcell" $environment)}}
+{{with $tcell := index $conf.Data}}
+
+{
+  "version": 1,
+  "applications": [
+    {
+      "host_identifier": "study server backend",
```
Comment (zyme) [inline pepper-apis/config/tcell_agent.config.ctmpl]: Would be great to set `host_identifier` as an env var so we can keep housekeeping and server straight in tcell dashboard.

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/StudyBuilderCli.java [hunk 2]
```diff
@@ -83,7 +83,7 @@ private void run(String[] args) throws Exception {
 
         Config cfg = ConfigFactory.load();
         TransactionWrapper.reset();
-        TransactionWrapper.init(new TransactionWrapper
+        TransactionWrapper.init("UTC", new TransactionWrapper
```
Comment (MocanaAtBroad) [inline study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/StudyBuilderCli.java]: Not needed.

**Review by MocanaAtBroad — COMMENTED** [review]
Silly question: At the end of day how do we know tcell is properly configured?

File: .circleci/config.yml [hunk 3]
```diff
@@ -53,6 +53,23 @@ executors:
     - image: *ci_support_image
 
 commands:
+  install-tcell:
+    steps:
+      - run:
+          name: Download tcell files from bucket
+          command: |
+            vault read -format=json  secret/pepper/prod/v1/conf | jq -r .data.gcp.serviceKey | gcloud auth activate-service-account --key-file=-
```
Comment (MocanaAtBroad) [inline .circleci/config.yml]:
```suggestion
    parameters:
      tcell_gcp_path:
        type: string
        default: *tcell_gcp_path
    steps:
      - run:
          name: Download tcell files from bucket
          command: |
            TCELL_PATH='<< parameters.tcell_gcp_path >>'
            vault read -format=json  secret/pepper/prod/v1/conf | jq -r .data.gcp.serviceKey | gcloud auth activate-service-account --key-file=-

```

File: .circleci/config.yml [hunk 4]
```diff
@@ -53,6 +53,23 @@ executors:
     - image: *ci_support_image
```
Comment (MocanaAtBroad) [inline .circleci/config.yml:56]: Add the following to the "references" section above (github won't let comment on that part of file):
```yaml
tcell_gcp_path: &tcell_gcp_path
            gs://ddp-tcell/tcell-1.10.0.tar.gz

File: .circleci/config.yml [hunk 5]
```diff
@@ -110,7 +131,9 @@ commands:
             esac
             echo "Setting deployment ENVIRONMENT to $DEPLOY_ENV"
             echo "export ENVIRONMENT=$DEPLOY_ENV" >> $BASH_ENV
-            source $BASH_ENV          
+            # to install a new version of tcell, do  "tar -cvfz tcell-x.y.z.tar.gz tcell" after expanding tcell's zip
+            echo 'export TCELL_PATH=gs://ddp-tcell/tcell-1.10.0.tar.gz' >> $BASH_ENV
```
Comment (MocanaAtBroad) [inline .circleci/config.yml]: Let's drop this line in favor of YAML reference that has string and command parameter that refers to the reference. Keeps the file path to something akin to a static header declaration.

File: .circleci/config.yml [hunk 6]
```diff
@@ -53,6 +53,23 @@ executors:
...
+
+            if [ -z $TAR_FILE_URL ]
+            then
+              echo "Could not find tcell archive ${TCELL_PATH}"
+              exit 1
+            fi
+            TAR_FILE_PATH="/tmp/$(basename -- $TAR_FILE_URL)"
+            gsutil cp  $TAR_FILE_URL $TAR_FILE_PATH
+
+            tar --no-same-owner -xvf $TAR_FILE_PATH
```
Comment (MocanaAtBroad) [inline .circleci/config.yml:80]: Mind adding a comment that you expect this command will create a directory called tcell ? Had to figure where the `tcell` directory was coming from in deploy and concluded it had to be from here.

File: .circleci/config.yml [hunk 7]
```diff
@@ -232,13 +255,16 @@ commands:
...
       - render-configs
+      - install-tcell
       - run:
           name: Gather deployment assets
           command: |
             DEPLOY_DIR='<<parameters.deploy_dir>>'
             rm -rf $DEPLOY_DIR
             mkdir -p $DEPLOY_DIR
             cd $DEPLOY_DIR
+            cp -r ../$TCELL_DIR .
```
Comment (MocanaAtBroad) [inline .circleci/config.yml:275]: Consider declaring `TCELL_DIR` in same way as the `DEPLOY_DIR` var is declared here.

File: pepper-apis/config/tcell_agent.config.ctmpl (hunk 1 above)
Comment (MocanaAtBroad) [reply pepper-apis/config/tcell_agent.config.ctmpl]: Was going to ask about confusing housekeeping and back-end but looks like you have concluded mixing the two is acceptable for now?

**Review by MocanaAtBroad — COMMENTED** [review]
Seeing some warnings and errors in GAE console that appear to be coming from TCELL. Here is a link to one of them:
https://console.cloud.google.com/logs/viewer?project=broad-ddp-dev&authuser=1&resource=gae_app%2Fmodule_id%2Fpepper-backend&minLogLevel=0&expandAll=false&timestamp=2020-04-27T18%3A56%3A30.710000000Z&customFacets&limitCustomFacetWidth=true&dateRangeStart=2020-04-27T17%3A56%3A21.252Z&interval=PT1H&logName=projects%2Fbroad-ddp-dev%2Flogs%2Fstdout&logName=projects%2Fbroad-ddp-dev%2Flogs%2Fstderr&logName=projects%2Fbroad-ddp-dev%2Flogs%2Fappengine.googleapis.com%252Fstdout&logName=projects%2Fbroad-ddp-dev%2Flogs%2Fappengine.googleapis.com%252Fstderr&logName=projects%2Fbroad-ddp-dev%2Flogs%2Fappengine.googleapis.com%252Frequest_log&scrollTimestamp=2020-04-27T18%3A53%3A36.279960000Z&filters=text%3Atcell&dateRangeUnbound=forwardInTime&advancedFilter=resource.type%3D%22gae_app%22%0Aresource.labels.zone%3D%22us17%22%0Aresource.labels.project_id%3D%22broad-ddp-dev%22%0Aresource.labels.version_id%3D%225f8685d%22%0Aresource.labels.module_id%3D%22pepper-backend%22%0Atimestamp%3D%222020-04-27T18%3A53%3A35.901266000Z%22%0AinsertId%3D%225ea72a2f000dc0925181b95b%22

File: .circleci/config.yml (hunk 7 above)
Comment (zyme) [reply .circleci/config.yml:275]: Nice, very elegant!

File: .circleci/config.yml (hunk 5 above)
Comment (MocanaAtBroad) [reply .circleci/config.yml]: Looks like you can get rid of this export now.

**Review by MocanaAtBroad — COMMENTED** [review]
Looks like we are still getting tcell warnings and errors at startup.
What I have been doing is going to https://pepper-backend-dot-broad-ddp-dev.uc.r.appspot.com/ to force the back-end to start (our GAE setup requires that a request come through), wait a couple of minutes, and then check the ddp-dev pepper-backend logs searching for "tcell"
Screen Shot 2020-04-28 at 2 28 51 PM


https://console.cloud.google.com/logs/viewer?project=broad-ddp-dev&authuser=1&resource=gae_app%2Fmodule_id%2Fpepper-backend&minLogLevel=0&expandAll=false&timestamp=2020-04-28T18%3A19%3A29.115000000Z&customFacets&limitCustomFacetWidth=true&dateRangeStart=2020-04-28T17%3A19%3A47.906Z&dateRangeEnd=2020-04-28T18%3A19%3A47.906Z&interval=PT1H&logName=projects%2Fbroad-ddp-dev%2Flogs%2Fstdout&logName=projects%2Fbroad-ddp-dev%2Flogs%2Fstderr&logName=projects%2Fbroad-ddp-dev%2Flogs%2Fappengine.googleapis.com%252Fstdout&logName=projects%2Fbroad-ddp-dev%2Flogs%2Fappengine.googleapis.com%252Fstderr&logName=projects%2Fbroad-ddp-dev%2Flogs%2Fappengine.googleapis.com%252Frequest_log&scrollTimestamp=2020-04-28T18%3A18%3A47.847208000Z&filters=text%3Atcell&advancedFilter=resource.type%3D%22gae_app%22%0Aresource.labels.version_id%3D%226b557db%22%0Aresource.labels.module_id%3D%22pepper-backend%22%0Aresource.labels.zone%3D%22us17%22%0Aresource.labels.project_id%3D%22broad-ddp-dev%22%0Atimestamp%3D%222020-04-28T18%3A18%3A47.688982000Z%22%0AinsertId%3D%225ea87387000a8356302949fb%22

File: pepper-apis/config/tcell_agent.config.ctmpl (hunk 1 above)
Comment (zyme) [reply pepper-apis/config/tcell_agent.config.ctmpl]: I'm going to try setting env vars and see if that does the trick...

**Comment by zyme** [comment]
@MocanaAtBroad thanks for trolling the logs--that cache policy error is something I'll bring back to tcell, since it's the same "read only filesystem" that led to their fixes last week.

**Comment by zyme** [comment]
Env vars for host names works!
Screen Shot 2020-04-28 at 3 27 54 PM

**Review by MocanaAtBroad — COMMENTED** [review]
Do we need to keep tcell jars and libraries in repo if we are going to grab them from a bucket at deploy time?

**Comment by MocanaAtBroad** [comment]
And if same stuff in repo, would it be preferable to make libraries part of mega jar by including them as Maven dependencies. There are some mechanisms available to include local libraries as Maven dependencies. https://devcenter.heroku.com/articles/local-maven-dependencies.

**Comment by zyme** [comment]
@MocanaAtBroad I've removed the tcell jars from source since we grab them from a bucket now.

**Review by MocanaAtBroad — DISMISSED** [review]
The code changes look good and work as intended.
I don't feel qualified to declare that Tcell is working correctly. The warnings and errors in the logs are not good. But if we are getting what we expect from the console...

**Review by MocanaAtBroad — COMMENTED** [review]
@zyme Could you do me a favor and remove the entry for : `build-executor` under the `executors` sections of `config.yml` ? It is unused and just don't want to go through the whole PR rigamarole to clean it up.

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**